Adding Hamiltonian Method for Equations of Motion

AnimeshSinha1309 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixed #16285 - Atleast Partially

#### Brief description of what is fixed or changed
Given the Hamiltonian, the position and momenta coordinates, the method sympy.calculus.hamilton.hamilton_equations returns the 2N first order equations of motion, where N is the number of Position or Momentum coordinates. 

#### Other comments
Method Signature: `hamilton_equations(Hamiltonian, [Position Functions], [Momentum Functions], Time Symbol)`

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Added function hamilton_equations to use Hamiltonian to get to the equations of motion
<!-- END RELEASE NOTES -->

AnimeshSinha1309 committed:
Starting out with the Hamiltonian

The most basic version possible that serves the purpose with a fixed time variable. Added to the Euler module.
Attempt at Fixing #16285.

AnimeshSinha1309 committed:
Files and Tests

Added one test for the hamiltonian function, moved it to a file of it's own. Fixed a sign issue.

oscargus commented on sympy/calculus/hamilton.py:
```diff
@@ -0,0 +1,75 @@
+
+def hamilton_equations(H, coords, momenta, time):
+    r"""
+    Find the Euler-Lagrange equations [1]_ for a given Lagrangian.
```
Should this read "Hamilton's"?

AnimeshSinha1309 replied:
I will definitely have to change the comment. But I kept it Hamilton instead of Hamilton's because in the other function it was Euler instead of Euler's. So just for consistency, tell me if I should make it Hamilton's.

smichr replied:
I think non-possessive is ok; in this case the name acts as an adjective as in "Hamiltonian mechanics".

smichr replied:
Would "Return the equations of the Hamiltonian..." work?

smichr replied:
or "Hamiltonian equations of motion".

AnimeshSinha1309 committed:
Adding Tests

Added more tests for Hamiltonian and changing the comments.

AnimeshSinha1309 committed:
Removed a Test

Test was not working on Python2 and PyPI.

AnimeshSinha1309 commented:
Okay, now everything works. Please Review.

oscargus commented:
Can you please add tests for the exceptions as well? Use `raises(ExpectedException, lambda: expression_that_triggers_exception)`. With that, the code will be fully covered.

AnimeshSinha1309 committed:
Added more tests

Improving test coverage specifically for error cases in the hamilton_equations method.

AnimeshSinha1309 commented:
@oscargus or anyone, please review, I added in the extra tests checking for possible error cases.

oscarbenjamin commented:
I don't think this belongs in calculus. Lagrange is in physics/mechanics. This should go somewhere round there.

AnimeshSinha1309 commented:
@oscarbenjamin, I agree that this might be a Physics thing, but the Generation of the Euler-Lagrange equations is in the `sympy.calculus.euler` module, in the same directory. The Lagrangian formulation in the Physics module seems to be completely different, its modelling a system and then getting the equations of motion.
Since it's a sister function (First Order Differential Equations instead of Second Order, but the same thing) to `euler.langrange_equations` in the same folder, I kept it there. Should I change something, Your thoughts?

AnimeshSinha1309 commented:
I know everyone is busy, but could someone please review this PR, and tell me if something needs to be done? @oscargus @smichr

AnimeshSinha1309 committed:
Merge branch 'master' into feature/hamiltonian-16285

oscarbenjamin commented on sympy/calculus/tests/test_hamilton.py:
```diff
@@ -0,0 +1,84 @@
+    t = Symbol('t')
+    h = x(t) ** 2 / 2 + p(t) ** 2 / 2
+    # Giving function instead of symbol as time
+    raises(TypeError, lambda: hamilton(h, [x(t)], [p(t)], x(t)))
```
These tests are quite repetitive. They should probably be combined into a single `test_` function.

AnimeshSinha1309 replied:
I merged the Error cases, but not the SHM because soon I want to write better tests for more complex systems and it's easier to visualise each system in it's own function rather than clubbed together.

oscarbenjamin commented on sympy/calculus/hamilton.py:
```diff
@@ -0,0 +1,75 @@
+
+def hamilton_equations(H, coords, momenta, time):
+    r"""
+    Find the Hamilton's equations [1]_ for a given Hamiltonian.
```
I don't think you need the citation reference here

AnimeshSinha1309 replied:
Yeah, didn't see that, got rid of it, thanks.

oscarbenjamin commented on sympy/calculus/hamilton.py:
```diff
@@ -0,0 +1,75 @@
+        List of conjugate momentum coordinates that the Hamiltonian depends upon.
+
+    time : Symbol
+        The Symbols used for time, here the only free variable in the equation.
```
Symbols should be singular

AnimeshSinha1309 replied:
Changed that too.

oscarbenjamin reviewed (commented):
It would be nice if this and the equivalent for Lagrangians could be used together somehow. Say I defined T and U and then wrote L=T-U and H=T+U and could get the Euler-Lagrange equations or Hamilton's equations for the same system. That doesn't seem directly possible here since the Hamiltonian has to use different symbols for the momenta whereas the Lagrangian is written terms of derivatives of the generalised coordinates. Is it possible to integrate these more?

AnimeshSinha1309 committed:
Doc issues, depreciation, tests

Minor issues with PR resolved.

AnimeshSinha1309 committed:
Merge branch 'feature/hamiltonian-16285' of github.com:AnimeshSinha1309/SympyRepo into feature/hamiltonian-16285

AnimeshSinha1309 committed:
Removed Whitespace

Was failing code quality check. Accidental whitespace removed.

AnimeshSinha1309 commented:
@oscargus I am really not sure how we can get Lagrangian and Hamiltonian in the same form, because at the end of the day there will be 2n free variables in Hamiltonian where as n in Lagrangian, and it seems unnecessary to auto collect the position and momenta coordinates using some convention we impose. I don't see a nice way to make it more uniform, but if you have something in mind, please do tell me and I will implement it.

oscarbenjamin commented:
I guess you mean me rather than @oscargus. Maybe there isn't a good way to combine the Hamiltonian and Lagrangian functions here.

AnimeshSinha1309 committed:
Tabs to Spaces again

Trying to fix the issue that tabs are showing up in the files.

AnimeshSinha1309 commented:
Yeah, sorry, @oscarbenjamin. Yeah, I don't see a better way to combine those two, this is the best I can manage for now, though having a more complex system class in the physics module would eventually work better. But this should be good for now. Please review and tell me if something more is needed.

smichr commented on sympy/calculus/tests/test_hamilton.py:
```diff
@@ -0,0 +1,87 @@
+    k = Symbol('k')
+    t = Symbol('t')
+    h = x(k) ** 2 / 2 + p(k) ** 2 / 2
+    # Giving function instead of symbol as time
```
```suggestion
    # Giving wrong dependent variable
```

AnimeshSinha1309 commented:
Does anything more need to be done?

AnimeshSinha1309 commented:
@oscarbenjamin @oscargus @smichr Can any of you please tell me if there is something more is to be done here, or approve this.

smichr commented on sympy/calculus/tests/test_hamilton.py:
```diff
@@ -0,0 +1,87 @@
+    pt = Function('pTheta')
+    pz = Function('pZ')
+    h3 = pt(t)**2/2 + pz(t)**2/2 + qz(t)**2/2
+    assert hamilton(h3, [qr(t), qt(t), qz(t)], [pr(t), pt(t), pz(t)], t) == [
```
I imagine that most people would create the Functions like symbols and then use them in an expression. We could eliminate a lot of redundancy by allowing entry like `(h3, (qr, qt, qz), (pr, pt, pz), t)` and then recompose them in the function: 
```python
for q, p in funcs:
    coord, momentum = q(t), p(t)
...
```

smichr replied:
(But if the arguments would have be more than `t` then this wouldn't work.)

AnimeshSinha1309 replied:
Yeah, but won't it just be better to leave it as is, since it's both similar to it's Lagrangian counterpart, and as you pointed out, could potentially have more arguments than t. Tell me what I should do @smichr.

AnimeshSinha1309 commented:
That was just a mess-up with another branch, I confused the issues. No changes were made to the code in the Forced push do and undo.

czgdp1807 commented:
@AnimeshSinha1309 It looks like a good work. Would you still like to continue your work on this? Please let us know.

AnimeshSinha1309 commented:
@czgdp1807 I would love to contribute on Physics and Quantum side of things, but to try Quantum I need some help, and last I heard there was no mentor there. Please tell me if there is something I can help out with.

czgdp1807 commented:
@oscarbenjamin @smichr  Can you please help(reviews, comments for completion) @AnimeshSinha1309 ? The work looks good but I am not much familiar with these modules of `SymPy`. 
@AnimeshSinha1309 For a start, please merge `master` into `feature/hamiltonian-16285`.

oscarbenjamin commented:
I can see why this PR was opened because it is analogous to the `euler_equations` function for getting the equations of motion from a Lagrangian. However the `euler_equations` function is not used anywhere in SymPy. The code that actually implements Lagrange's method in `sympy.physics.mechanics.lagrange` has its own more complicated method for finding the equations of motion that also handles constraints etc so it has no need to use the `euler_equations` function.

I am not sure that it is useful to have a function for differentiating the Hamiltonian to get the equations of motion. If someone knows about Hamiltonian systems then they will know these equations and it will be straight-forward to do what this function does. Something like this would be useful if it linked up with other things like code for generating the Hamiltonian itself or for doing other things with Hamiltonians.

AnimeshSinha1309 commented:
Yeah, when I wrote this, I did because the Robot Arm Dynamics things had a lot to do with Hamiltonian, but it soon got too complicated for me. I still believe that this is a useful method by itself, even though not crucial, and the final implementation will depend heavily on what those Hamiltonian Dynamics methods need. Till that is done, of course some one can easily do this themselves, but it's also convenient and cleaner if I have it. If it's not, please do close this PR.

czgdp1807 commented:
> I am not sure that it is useful to have a function for differentiating the Hamiltonian to get the equations of motion. If someone knows about Hamiltonian systems then they will know these equations and it will be straight-forward to do what this function does. Something like this would be useful if it linked up with other things like code for generating the Hamiltonian itself or for doing other things with Hamiltonians.

ping @sympy/mechanics

czgdp1807 commented:
Closing this for now. However, if you think this should be added then please let us know in the comments below. Thanks for the contributions.
